provider/datagen/src/bin/icu4x-datagen.rs

```diff
@@ -311,55 +336,125 @@ fn main() -> eyre::Result<()> {
             .unwrap()
     }
 
-    let mut provider = DatagenProvider::new_custom();
+    #[allow(unused_mut)]
+    let mut preprocessed_locales = if cli.locales.as_slice() == ["none"] {
+        Some(PreprocessedLocales::LanguageIdentifiers(vec![]))
+    } else if cli.locales.as_slice() == ["full"] || cli.all_locales {
+        Some(PreprocessedLocales::All)
+    } else {
+        if cli.locales.as_slice() == ["all"] {
+            log::warn!(
+                "`--locales all` selects the Allar language. Use `--locales full` for all locales"
+            );
+        }
+        None
+    };
 
-    provider = provider.with_collation_han_database(match cli.collation_han_database {
-        CollationHanDatabase::Unihan => icu_datagen::CollationHanDatabase::Unihan,
-        CollationHanDatabase::Implicit => icu_datagen::CollationHanDatabase::Implicit,
-    });
+    let provider: Box<dyn ExportableProvider> = match () {
+        #[cfg(all(feature = "blob_input", not(feature = "provider")))]
+        () => Box::new(ReexportableBlobDataProvider(
+            icu_provider_blob::BlobDataProvider::try_new_from_blob(
+                std::fs::read(cli.input_blob)?.into(),
+            )?,
+        )),
+        #[cfg(all(feature = "blob_input", feature = "provider"))]
+        () if cli.input_blob.is_some() => Box::new(ReexportableBlobDataProvider(
+            icu_provider_blob::BlobDataProvider::try_new_from_blob(
+                std::fs::read(cli.input_blob.unwrap())?.into(),
+            )?,
+        )),
```

Comment:
`#[cfg(not(...))]` should almost always be avoided... please just write this as something more like

```rust
if let Some(input_blob_path) = cli.input_blob {
    // generate blob input
} else if cfg!(feature = "provider") {
    // generate DatagenProvider input
} else {
    eyre::bail!("...")
}
```

Reply:
doesn't work, `cfg!` evaluates to a boolean, it doesn't remove the branch from the code, so it won't compile because `DatagenProvider` isn't there.

Reply:
There are lots of `#[cfg(not(...))]` in this file. Unlike in the libraries we *want* different behaviour if a feature is disabled, i.e. to fall back to something or to explain what's wrong.

Reply:
Sure, good point, but every `#[cfg(not(...))]` resolves to an `eyre::bail!`, which means there's not any code paths that aren't being tested; they just resolve to an error case. I'm concerned about `#[cfg(not(...))]` that resolve to real code.

Reply:
I think this discussion is obsolete